src/v/storage/log_reader.cc

```diff
@@ -62,9 +62,9 @@ batch_consumer::consume_result skipping_consumer::accept_batch_start(
         _reader._config.start_offset = header.last_offset() + model::offset(1);
         return batch_consumer::consume_result::skip_batch;
     }
-    if (_reader._config.first_timestamp > header.first_timestamp) {
-        // kakfa needs to guarantee that the returned record is >=
-        // first_timestamp
+    if (_reader._config.first_timestamp > header.max_timestamp) {
+        // kakfa requires that we return messages >= the timestamp, it is
+        // permitted to include a few earlier
```

Comment:
well this change definitely looks correct to me, but it also seems so obvious that i wonder how the mistake was made in three separate places.

it may be related to limited testing and what looks problematic in our record batch builder where the first/max are set to the same timestamp:

https://github.com/redpanda-data/redpanda/blob/dev/src/v/storage/record_batch_builder.cc#L42-L43

the batch builder is used in testing, but, i think it is also used for internal batches, and the place that might be concerning is for compressed + compaction where i was thinking that we destroy and re-recreate the batch but i dunno if that applies to the header.

Reply:
As far as I can tell, the existing code was all just predicated on batches having equal first/max timestamps.

The way we build batches on compaction is interesting + I hadn't thought about it, I will make a note to get into that.

Reply:
> As far as I can tell, the existing code was all just predicated on batches having equal first/max timestamps.

or that all the batches redpanda was constructing were assumed to have 1 record.